src/EditorFeatures/Core/DefinitionContextTracker.cs

```diff
@@ -0,0 +1,211 @@
+            _codeDefinitionWindowService = codeDefinitionWindowService;
+        }
+
+        void IWpfTextViewConnectionListener.SubjectBuffersConnected(IWpfTextView textView, ConnectionReason reason, Collection<ITextBuffer> subjectBuffers)
+        {
+            if (!_views.ContainsKey(textView) && !textView.Roles.Contains(PredefinedTextViewRoles.CodeDefinitionView))
+            {
+                _views.Add(textView, new CancellationTokenSource());
+                textView.Caret.PositionChanged += OnTextViewCaretPositionChanged;
+                var fireAndForget = UpdateDefinitionContext(textView, textView.Caret.Position);
+            }
+        }
+
+        void IWpfTextViewConnectionListener.SubjectBuffersDisconnected(IWpfTextView textView, ConnectionReason reason, Collection<ITextBuffer> subjectBuffers)
+        {
+            if (reason == ConnectionReason.TextViewLifetime ||
+                !textView.BufferGraph.GetTextBuffers(b => b.ContentType.IsOfType(ContentTypeNames.RoslynContentType)).Any())
+            {
+                CancellationTokenSource cancellationTokenSource;
+                if (_views.TryGetValue(textView, out cancellationTokenSource))
+                {
+                    cancellationTokenSource.Cancel();
+                    _views.Remove(textView);
+                    textView.Caret.PositionChanged -= OnTextViewCaretPositionChanged;
+                }
+            }
+        }
+
+        private void OnTextViewCaretPositionChanged(object sender, CaretPositionChangedEventArgs e)
+        {
+            var fireAndForget = UpdateDefinitionContext(e.TextView, e.NewPosition);
+        }
+
+        private async Task UpdateDefinitionContext(ITextView textView, CaretPosition caretPosition)
+        {
+            // Cancel any pending update for this view
+            _views[textView].Cancel();
+
+            // See if we moved somewhere else in a projection that we care about
+            var pointInRoslynSnapshot = caretPosition.Point.GetPoint(tb => tb.ContentType.IsOfType(ContentTypeNames.RoslynContentType), caretPosition.Affinity);
+            if (pointInRoslynSnapshot == null)
+            {
+                return;
+            }
+
+            // After a delay in case the caret moves again, find the symbol under the caret and update the context
+            var cancellationTokenSource = new CancellationTokenSource();
+            _views[textView] = cancellationTokenSource;
+            var cancellationToken = cancellationTokenSource.Token;
+
+            var foregroundTaskScheduler = TaskScheduler.FromCurrentSynchronizationContext();
+
+            var locations = await Task.Run(
+                async () => await GetContextFromPoint(pointInRoslynSnapshot, foregroundTaskScheduler, cancellationToken).ConfigureAwait(false),
+                cancellationToken).ConfigureAwait(true);
+
+            _codeDefinitionWindowService.SetContext(locations);
```

Comment:
Do you need to recheck the cancellation token here, since it's possible that this operation was canceled while we redispatched to the UI thread?